libs/plugins/flow-client/src/lib/core/models/flow/insert-node.ts

```diff
@@ -1,28 +1,62 @@
-import { FlowGraph, GraphNode } from '@flogo-web/lib-client/core';
+import { FlowGraph, GraphNode, NodeType } from '@flogo-web/lib-client/core';
+import { BRANCH_PREFIX } from '@flogo-web/lib-client/diagram';
 
 export function insertNode(
   flowGraph: FlowGraph,
   node: GraphNode,
-  parentId?: string
+  parentId?: string,
+  insertBetween?: boolean
 ): FlowGraph {
-  const parent = flowGraph.nodes[parentId];
+  let parent = flowGraph.nodes[parentId];
   let nodes = flowGraph.nodes;
+  let rootId = flowGraph.rootId;
   if (parent) {
+    if (insertBetween) {
+      const nonBranchChild = getNonBranchChild(parent);
+      node = {
+        ...node,
+        children: [...nonBranchChild],
+      };
+      parent = { ...parent, children: [...getBranchChildren(parent), node.id] };
+    } else {
+      parent = { ...parent, children: [...parent.children, node.id] };
+    }
     nodes = {
       ...nodes,
-      [parent.id]: { ...parent, children: [...parent.children, node.id] },
+      [parent.id]: { ...parent },
     };
   } else {
-    flowGraph = {
-      ...flowGraph,
-      rootId: node.id,
-    };
+    // case when adding a tile before the root tile of the diagram
+    if (insertBetween) {
```

Comment:
like in my previous comment, I think we don't need this flag. Since at this point we know we want to insert in the root position because we know `parent==null` then we can depend on `flowGraph.rootId !== null` to know if this is an insert in between operation

```ts
if (parent) {
} else {
   const insertBeforeExistingRoot = flowGraph.rootId != null;
}
```